src/io/platform/f4xx/I2Cf4xx.cpp

```diff
@@ -0,0 +1,264 @@
+ * documentation with easier documentation on MBed's STM32F302r8
+ * documentation.
+ *
+ * TODO: Make modifications for this to be runnable at compile time
+ * allowing for validation of the selected pins.
+ *
+ * @param sclPin The selected I2C clock pin
+ * @return The port ID associated with the selected pins
+ */
+static uint8_t getPortID(Pin sclPin) {
+#ifdef STM32F302x8
+
+    switch (sclPin) {
+    case Pin::PB_8:
+    case Pin::PB_6:
+    case Pin::PA_15:
+        return 1;
+    case Pin::PA_9:
+    case Pin::PF_1:
+        return 2;
+    case Pin::PA_8:
+        return 3;
+    default:
+        log::LOGGER.log(log::Logger::LogLevel::ERROR, "Invalid SCL Pin");
+        return 0;
+    }
+#endif
+
+#ifdef STM32F334x8
+    return 1;
+#endif
+
+#ifdef STM32F4xx
+    switch (sclPin) {
+    case Pin::PB_8:
+        return 1;
+    case Pin::PB_10:
+        return 2;
+    default:
+        //        log::LOGGER.log(log::Logger::LogLevel::ERROR, "Invalid SCL Pin");
+        return 0;
+    }
+#endif
+}
+
+/**
+ * Get the target I2C instance based on the version of the STM32f3
+ *
+ * The STM32F302r8 supports 3 I2C instances, the STM32F334x8 supports
+ * 1 I2C instance
+ *
+ * @param portId The target port, for the STM32f334 this is always 1
+ */
+static void getInstance(uint8_t portID, I2C_TypeDef** instance, uint8_t* altId) {
+#ifdef STM32F302x8
+    switch (portID) {
+    case 1:
+        *instance = I2C1;
+
+        if (!__HAL_RCC_I2C1_IS_CLK_ENABLED())
+            __HAL_RCC_I2C1_CLK_ENABLE();
+
+        *altId = GPIO_AF4_I2C1;
+
+        break;
+    case 2:
+        *instance = I2C2;
+
+        if (!__HAL_RCC_I2C2_IS_CLK_ENABLED())
+            __HAL_RCC_I2C2_CLK_ENABLE();
+
+        *altId = GPIO_AF4_I2C2;
+
+        break;
+    case 3:
+        *instance = I2C3;
+
+        if (!__HAL_RCC_I2C3_IS_CLK_ENABLED())
+            __HAL_RCC_I2C3_CLK_ENABLE();
+
+        *altId = GPIO_AF2_I2C3;
+        break;
+    default:
+        break;
+    }
+#endif
+
+#ifdef STM32F334x8
+    *instance = I2C1;
+    *altId = GPIO_AF4_I2C1;
+    if (!(__HAL_RCC_I2C1_IS_CLK_ENABLED()))
+        __HAL_RCC_I2C1_CLK_ENABLE();
+#endif
+
+#ifdef STM32F4xx
+    switch (portID) {
+    case 1:
+        *instance = I2C1;
+
+        if (!__HAL_RCC_I2C1_IS_CLK_ENABLED())
+            __HAL_RCC_I2C1_CLK_ENABLE();
+
+        *altId = GPIO_AF4_I2C1;
+        break;
+    case 2:
+        *instance = I2C2;
+
+        if (!__HAL_RCC_I2C2_IS_CLK_ENABLED())
+            __HAL_RCC_I2C2_CLK_ENABLE();
+
+        *altId = GPIO_AF4_I2C2;
+
+        break;
+    default:
+        break;
+    }
+#endif
+}
+
+I2Cf4xx::I2Cf4xx(Pin sclPin, Pin sdaPin) : I2C(sclPin, sdaPin) {
+    GPIO_InitTypeDef gpioInit;
+    uint8_t altId = 0x00U;
+
+    uint8_t portId = getPortID(sclPin);
+    getInstance(portId, &halI2C.Instance, &altId);
+
+    Pin i2cPins[2] = {sclPin, sdaPin};
+    uint8_t numOfPins = 2;
+
+    //TODO: CHECK GPIO STATES IN HAL CODE
```

Comment:
Did you handle these TODOs?

Reply:
I checked the clock speed nad it is correct. Need to check the GPIO state